### DISCUSSION TITLE
Failed to connect with Adafruit MQTT

### DISCUSSION THREAD

**Discussion by I-Code0** [arendst/Tasmota#18324] [Q&A]
Hello Guys
I wondered if you could help me to fix the problem with connecting Tasmota to Adafruit MQTT.
The problem is MQTT ERROR: DVES_20AC64 31.56.199.250 SUBSCRIBE to topic `sensor/cmnd/#` rejected, not a valid resource topic
The User and Passwords are all corect and i see this error on my adafruit dashboard!
Tasmota console Error 12:15:55.916 MQT: Attempting connection...
12:15:56.130 MQT: Connect failed to io.adafruit.com:1883, rc -2. Retry in 10 sec

**Comment by sfromis** [comment]
Looks like Adafruit may have some strict requirements for how topics are constructed, and what you post about using may not match that.

What you configure in Tasmota full topic controls what Tasmota will subscribe to, and you may have to do a specific config of that.

**Reply by I-Code0** [reply]
I checked them ALL. they are all correct! exept the 'cmnd' at the last! how can remove it in Tasmota?
could you ever connect tasmota to adafruit?

**Reply by barbudor** [reply]
It's impossible to use AdafruitIO with Tasmota due to both have strict requirements on MQTT topics
You can't remove cmnd and other topics from Tasmota as Tasmota won't work without

The only solution could be to write dedicated AdafruitIO code into Tasmota MQTT module like it was done for AWS for example

Not worth it
There are free true MQTT broker available on the internet that works fine with Tasmota such as flespi.io or MyMqttHub or others

**Reply by I-Code0** [reply]
thank you so much @barbudor .
does any of them working with IFTTT or Zapier?
Actually I wanna send notification to my mobile whenever the pir sensor triggered.

**Reply by barbudor** [reply]
Tasmota can directly call an Iffft webhook using the command WebQuery. Don't even need a MQTT broker
Probably same with zappier

**Reply by I-Code0** [reply]
@barbudor thank you very much for your reply.
do you have any tutorial to make this happen?

**Reply by pkkrusty** [reply]
Could also do a publish to ntfy.sh. Not secure, but dead simple.

**Reply by sfromis** [reply]
Same issue asked here: https://discord.com/channels/479389167382691863/479389167382691865/1092553693250990120

**Reply by I-Code0** [reply]
I just checked it and in some line I get an error for connection which is not important that much but in setting the rules to trigger the webhooks I wasn't successful with the this rule:
`rule1 on switch1#state=1 do WebSend [MY NTFY LINK] endon`
do you have any suggestion?

**Reply by sfromis** [reply]
If what you have for the link is an URL, do not use `WebSend` but instead `WebQuery` (and remember the `GET` command).

**Reply by I-Code0** [reply]
> If what you have for the link is an URL, do not use `WebSend` but instead `WebQuery` (and remember the `GET` command).

changed the rule to 
`Rule1 ON Switch1#state=1 do WebQuery [http://ntfy.sh/topic POST [Title: Motion Sensor] Activated] [GET] endon`
but nothing is triggered,which part of the code is wrong?

**Reply by sfromis** [reply]
Instead of that weird contraption of sprinkling `[]` all over, do read docs and follow the syntax of the command.

And of course, make sure that the rule is activated too.

**Reply by I-Code0** [reply]
The Rule model is created by ChatGBT.
I searched a lot and there is no sample model of the rule in docs for WebQuery!

**Reply by sfromis** [reply]
> The Rule model is created by ChatGBT.

This means that it is *utter garbage*. For Tasmota, ChatGPT is much worse than it's common incompetency for details.

You did really not search much if you did not find the link to "more information" in the docs entry for the command [WebQuery](https://tasmota.github.io/docs/Commands/#webquery).

**Reply by I-Code0** [reply]
> > The Rule model is created by ChatGBT.
> 
> This means that it is _utter garbage_. For Tasmota, ChatGPT is much worse than it's common incompetency for details.
> 
> You did really not search much if you did not find the link to "more information" in the docs entry for the command [WebQuery](https://tasmota.github.io/docs/Commands/#webquery).
Dear @sfromis I have read them all and check all the command!
then asked from Chat gbt which gave me the wrong code.
one of the rules I made by the link you gave is this.
`Rule1 ON Switch1#State=1 DO WebQuery http://ntfy.sh/pfoe POST [Title: Dishwasher] {Program finished} GET ENDON`
I just wanna sure that the rule is correct or not! maybe the problem is from my network side or something else!

**Reply by sfromis** [reply]
While that rule is "close" to right syntax (assuming that ntfy expects Title as a http header field), all that you have after the header fields `[]` is message body, and as you have the command of POST (instead of GET) placed where it should be (right after the URL), you should likely not also have that GET at the end, unless you want that as part of the message body.

Your usage of `{}` is "unusual", unless you specifically want those chars in the payload (it is not part of the WebQuery syntax). Typically, `{}` is used with JSON, but your payload is not JSON, unlike the example in the thread.

Of course, you should also make sure that your ntfy request meets their requirements exactly.

**Reply by I-Code0** [reply]
> While that rule is "close" to right syntax (assuming that ntfy expects Title as a http header field), all that you have after the header fields `[]` is message body, and as you have the command of POST (instead of GET) placed where it should be (right after the URL), you should likely not also have that GET at the end, unless you want that as part of the message body.
> 
> Your usage of `{}` is "unusual", unless you specifically want those chars in the payload (it is not part of the WebQuery syntax). Typically, `{}` is used with JSON, but your payload is not JSON, unlike the example in the thread.
> 
> Of course, you should also make sure that your ntfy request meets their requirements exactly.

Thank you very much for you kind reply.
Is there any possibility to check that what's happening when the switch1 is ON in console?
Because I don't see the command triggering in console while the PIR sensor trigger, it is only `POWER = ON`
Everything seems to be alright but nothing work.

**Reply by sfromis** [reply]
Did you make sure that the rule is activated? Not just saved.... And have a SwitchMode to match....

Yes, you can check what's happening by watching the console. Do activate `Weblog 3` for additional details.

**Comment by I-Code0** [comment]
The rule is activated with rule1 1
please see the attached files
![tasmota](https://user-images.githubusercontent.com/87815260/229909631-2d945199-21ba-4233-acd2-cb103e587ec6.jpg)
[tasmo.txt](https://github.com/arendst/Tasmota/files/11152668/tasmo.txt)

**Reply by pkkrusty** [reply]
Here's a working webquery that I use with one of my devices. Not sure if you have that part figured out yet. `webquery http://ntfy.sh/pkkrusty-example POST [Title: Water Heater] %var1%`

![IMG_3409097A7418-1](https://user-images.githubusercontent.com/79770016/229914939-3cdb3170-814a-444f-9e37-2a10ea4f79e4.jpeg)

When I subscribe to your topic on ntfy, this is what I see.  So you are getting something through.

**Reply by I-Code0** [reply]
@pkkrusty Thank you so much.
do you have some sample rule code like mine? if switch1 turned on ,send webquery to ntfy

**Reply by pkkrusty** [reply]
Here's my somewhat complex rule to send me a notification if my water heater gets turned on after 5pm:
```
rule1
ON Power1#State DO IF ((%value%==1) and (time>1020)) backlog event notify; var1 On after 5.; var2 3 ENDIF ENDON 
on event#notify do backlog webquery http://ntfy.sh/pkkrusty-example POST [Title: Water Heater] %var1% endon 
on webquery#Data$|failed do if (%var2%>0) backlog sub2 1; event notify endif endon
```
On a power state change, (I use if/expressions compiled in) I call an event. The event does the webquery. Then I have a trigger when the webquery result comes in (note that I wrap the webquery in backlog so I can access that trigger) to see if the POST failed or not.  If it was successful, no further action. If it failed, I try three more times. ( added the retries, because I noticed the webqueries were failing (maybe the timeout was to short for my crappy internet connection))

I don't think I've ever triggered on a switch#state. How is your dishwasher sending a 3.3v signal to your device?

**Reply by I-Code0** [reply]
@pkkrusty In my case I can send webquries by `webquery http://ntfy.sh/pfoe POST [Title: motion]`  
BUT when I try to use it as rule it doesn't work!
It must be something easy but I had no luck to fix it!

**Reply by pkkrusty** [reply]
From your log, it looks like switch1#state=1 is not triggering. Something isn't working with that piece. Maybe needs to be switch1#state=ON *edit* according to docs, I am wrong.  Needs to be as you have it entered.

Easiest solution is to just define a dummy relay, and trigger off of power1#state=1.

**Reply by sfromis** [reply]
Depends on what SwitchMode you have.

**Reply by I-Code0** [reply]
> power1#state=1
Dear @sfromis is that possible for you to make sample rule to check that it is the problem from rule code or something else?
send WebQuery to the link {LINK} When Switch1 trigger.

**Reply by sfromis** [reply]
Again, select a `SwitchMode` to match what you want to do.

You can check what state values you get using a test rule like `on Switch1#state do var1 %value% endon`

**Reply by I-Code0** [reply]
> Again, select a `SwitchMode` to match what you want to do.
> 
> You can check what state values you get using a test rule like `on Switch1#state do var1 %value% endon`

This is what I get `RESULT = {"Var1":"2"}`

**Reply by I-Code0** [reply]
The Switch1 Get value 2 instead of 1 and I had to change the value to 2.
but there is another problem which is from Internet side,Sometimes the webqueries Failed to send.is there any solution for that to be sure the Query send is done?

**Reply by sfromis** [reply]
Or... again ... select a `SwitchMode` to match what you want to have.

If you want to "be sure", you can look at the result of the command.

**Reply by pkkrusty** [reply]
@I-Code0 See my post up above to make the webquery try multiple times until successful.
```
rule1
ON Power1#State DO IF ((%value%==1) and (time>1020)) backlog event notify; var1 On after 5.; var2 3 ENDIF ENDON 
on event#notify do backlog webquery http://ntfy.sh/pkkrusty-example POST [Title: Water Heater] %var1% endon 
on webquery#Data$|failed do if (%var2%>0) backlog sub2 1; event notify endif endon
```
